### ISSUE TITLE
【请教/咨询】关于字级检错门控模型 ChineseErrorDetectorElectra 训练源码与细节的咨询

### ISSUE THREAD

**Issue by crisyy** [TW-NLP/ChineseErrorCorrector#49]
您好！

非常感谢贵团队开源了如此优秀的中文文本纠错平台 ChineseErrorCorrector，项目内容非常丰富且实用，给我们的学习和研究带来了很大的启发。

在关注到项目使用的字级检错门控模型 `xurong123/ChineseErrorDetectorElectra`后，我对其作为大模型纠错前置门控的设计非常感兴趣，也正在尝试在本地复现和搭建类似的轻量级 ELECTRA 字级检测 Pipeline。

目前，我在本地尝试按 README 的思路复现字级 TokenClassification 训练，但复现出来的模型效果明显达不到README上的指标，错字 召回（检出率）偏低，远低于 README 中 spelling recall ≈ 0.986 的水平，检出质量较差，误报/漏报偏多。

我想核对自己这套流程和原模型实现的差距，从仓库我已经能看到一部分配置，但真正影响复现的几块代码/配置仓库里没有，想请教：

您有没有 ChineseErrorDetectorElectra  公开训练/评测源码？尤其是：
``` 
char_align.py：source/target 的字级对齐打标逻辑（difflib.SequenceMatcher 的 replace/delete/insert 具体怎么落标）——这块差一点，标签就全偏了，最影响检出率；

train_char.py / CharTokenTrainer：加权 CE 的具体实现、subword→字的 offset 对齐、子词内多个 token 的标签如何归并；

evaluate_char.py：指标是按字还是按 token、是否过滤特殊 token / 子词非首位，这会直接改变 recall 数字。
```

主要目的是核对我本地训练的正确性，不会用于其它用途。非常感谢！！

**Comment by Shixurong** [comment]
感谢详细反馈。先说明一下现状：目前仓库公开的是推理侧代码；字级 TokenClassification 的完整训练/评测脚本尚未全部开源，所以 README 里的指标若按「自己从头训一版」去对齐，确实容易在几处细节上产生偏差。下面把真正影响 recall 的关键规则用文字 + 示例代码说明；若你需要可对齐 README 的完整 train_char.py / evaluate_char.py / char_align.py，可以留言，我可以单独提供（或按你的复现场景给一份最小可跑版本）。

1. README 里 spelling recall ≈ 0.986 是什么口径？
该数字来自分层验证集上 task=spelling 子集的字级（character-level） binary recall，不是：

任意验证集全量
subword token 粒度直接算 P/R
句级 max(p_ERR) >= 0.5 这类 gate 阈值下的数字
若验证集、打标、评测粒度任一处不一致，recall 会明显低于 README，这比较常见。感谢详细反馈。先说明一下现状：目前仓库公开的是推理侧代码；字级 TokenClassification 的完整训练/评测脚本尚未全部开源，所以 README 里的指标若按「自己从头训一版」去对齐，确实容易在几处细节上产生偏差。下面把真正影响 recall 的关键规则用文字 + 示例代码说明；若你需要可对齐 README 的完整 train_char.py / evaluate_char.py / char_align.py，可以留言，我可以单独提供（或按你的复现场景给一份最小可跑版本）。

1. README 里 spelling recall ≈ 0.986 是什么口径？
该数字来自分层验证集上 task=spelling 子集的字级（character-level） binary recall，不是：

任意验证集全量
subword token 粒度直接算 P/R
句级 max(p_ERR) >= 0.5 这类 gate 阈值下的数字
若验证集、打标、评测粒度任一处不一致，recall 会明显低于 README，这比较常见。

2. 字级打标（char_align 核心逻辑）
训练/评测前，source、target 需走同一套归一化（如 NFKC + 空白折叠）。对齐规则如下：

SequenceMatcher(..., autojunk=False)
replace / delete：source 侧对应区间标 1
insert：只发生在 target 侧，不占 source 下标，source 不标
标签长度 = len(source)
示例（与实现一致的最小版）：


import difflib
def char_error_labels(source: str, target: str) -> list[int]:
    labels = [0] * len(source)
    matcher = difflib.SequenceMatcher(a=source, b=target, autojunk=False)
    for tag, i1, i2, _j1, _j2 in matcher.get_opcodes():
        if tag == "equal":
            continue
        if tag in ("replace", "delete"):
            for i in range(i1, i2):
                labels[i] = 1
        # insert: 不标 source
    return labels


3. subword → 字：训练时的标签对齐
必须使用 Fast Tokenizer（支持 return_offsets_mapping）。规则：

[CLS] / [SEP] / padding：(0, 0) → label -100（不参与 loss）
普通 token 覆盖 [start, end)：取 char_labels[start:end] 的 max（有错即 ERR）
示例：

label_ids = []
for start, end in offset_mapping:
    if start == 0 and end == 0:
        label_ids.append(-100)
    else:
        seg = char_labels[start:min(end, len(char_labels))]
        label_ids.append(max(seg) if seg else 0)
推理还原到字级时，与训练对称：任一覆盖该字的 subword 预测 ERR，则该字为 ERR；p(ERR) 取覆盖该字各 token 的 max。

4. 损失函数（影响 recall 的常见训练侧差异）
README 基线若只写了「加权 CE + recall 选模」，复现时还可能涉及（具体超参以当时实验为准）：

class weight：[1, w_err]，ERR 类可乘 err_weight_mult
可选 Focal Loss、neg_subsample_keep（OK token 子采样，ERR 全保留），缓解负类淹没导致 recall 上不去
若你本地是「纯 CE + 默认权重 + 随机切 val」，和 README 数字不对齐是可能的。

5. 评测口径（evaluate_char 要点）
字级 P/R/F1：把所有句子的字摊平后，average="binary"（positive = 错字）
不算 special token；只统计 tokenizer 截断后覆盖到的字区间（gold/pred 对齐同一 char_end）
--by_task：在 val_task_stratified.jsonl 上按 task 分组；spelling recall ≈ 0.986 对应 task=spelling 那一行
--sweep_sentence 是句级 gate，不改变字级 recall

是否需要我提供完整训练/评测源码？
若你希望我直接发 char_align.py / prepare_char_jsonl.py / train_char.py / evaluate_char.py 的可运行版本

**Comment by crisyy** [comment]
非常感谢您的详细解答 🙏 
我对照了您说的规则，几处口径我之前确实没对齐，尤其是 insert 不在 source 标注、subword→字取 max、以及评测是把字摊平后按 average="binary" 算、不计 special token。我先按这几条规则改本地实现，很可能我之前问题就出在标注和评测口径上。

另外想顺便确认几个 训练侧超参：
学习率 / epoch 数 / batch size / warmup / weight decay  以及随机种子；

最后，还是希望麻烦您能发一份完整可跑版：char_align.py / prepare_char_jsonl.py / train_char.py / evaluate_char.py

再次感谢！！

**Comment by Shixurong** [comment]
> 非常感谢您的详细解答 🙏 我对照了您说的规则，几处口径我之前确实没对齐，尤其是 insert 不在 source 标注、subword→字取 max、以及评测是把字摊平后按 average="binary" 算、不计 special token。我先按这几条规则改本地实现，很可能我之前问题就出在标注和评测口径上。
> 
> 另外想顺便确认几个 训练侧超参： 学习率 / epoch 数 / batch size / warmup / weight decay 以及随机种子；
> 
> 最后，还是希望麻烦您能发一份完整可跑版：char_align.py / prepare_char_jsonl.py / train_char.py / evaluate_char.py
> 
> 再次感谢！！

[chinese_sentence_error_detector_char_scripts.tar.gz](https://github.com/user-attachments/files/28993313/chinese_sentence_error_detector_char_scripts.tar.gz)

**Comment by crisyy** [comment]
非常感谢，源码已收到 🙏 

我会按您这套配方重新跑一版来对齐 README 指标，有结果再同步。

再次感谢您的开源和耐心解答！！！

**Comment by crisyy** [comment]
您好：

   我尝试做了复现，但是字级 ERR Recall 与发布模型存在约 15 个点差距，不知是否和训练数据还有关系？

### 训练配置（对齐 training_char_meta.json）

| 参数 | 本次训练 | 原始模型 |
|---|---|---|
| base model | chinese-electra-180g-base-discriminator | 相同 |
| loss_type | ce | ce |
| err_weight_mult | 2.0 | 2.0 |
| final_class_weights[ERR] | 11.529 | 11.573 |
| val_ratio | 0.05 | 0.05 |
| train_n | 1,908,531 | 1,908,150 |
| val_n | 100,450 | 100,429 |
| metric_for_best_model | eval_recall | eval_recall |
| epochs | 3 | 3 |
| batch_size | 8 | 8 |
| lr | 2e-5 | 2e-5 |
| seed | 42 | 42 |
| precision | fp32 | fp32（推测） |

---

### 评测结果对比

评测集：`val_task_stratified_char.jsonl`（7500 条，spelling / grammar / mixed 各 2500）

#### 字级整体

| 指标 | 本次训练 | 原始模型 | 差值 |
|---|---|---|---|
| Accuracy | 0.9461 | 0.9081 | +0.038 |
| Precision (ERR) | 0.3637 | 0.2562 | +0.107 |
| **Recall (ERR)** | **0.6745** | **0.8222** | **-0.148 ↓** |
| F1 (ERR) | 0.4726 | 0.3906 | +0.082 |

#### 字级按任务拆分

| Task | 指标 | 本次训练 | 原始模型 | 差值 |
|---|---|---|---|---|
| spelling | Recall | 0.979 | 0.986 | -0.007 |
| grammar | Recall | 0.683 | 0.823 | -0.140 ↓ |
| mixed | Recall | 0.501 | 0.732 | -0.231 ↓ |

#### 句级（最优阈值）

| 指标 | 本次训练（thr=0.65） | 原始模型（thr=0.70） |
|---|---|---|
| Precision | 81.8% | 78.7% |
| Recall | 91.0% | 94.6% |
| F1 | **86.1%** | **85.9%** |

句级 F1 基本持平（差 0.2%），但字级 ERR Recall 差距集中在 grammar 和 mixed 两类。

---

### 问题

我目前使用的是 `train_stage1.json` + `train_stage2.json` 直接合并得到的 `train_stage_all.jsonl`（共 2,008,981 条，训练集 1,908,531 条）。

请问原始训练使用的 `cec_train.jsonl` 是如何从 `train_stage1/train_stage2` 生成的？

具体想确认：
1. 训练集是否经过了额外的清洗、去重或过滤步骤？
2. 生成脚本是否可以公开？
3. 请问是否我还有其他遗漏的超参？


非常感谢！！！

**Comment by Shixurong** [comment]
> 您好：
> 
> 我尝试做了复现，但是字级 ERR Recall 与发布模型存在约 15 个点差距，不知是否和训练数据还有关系？
> 
> ### 训练配置（对齐 training_char_meta.json）
> 参数	本次训练	原始模型
> base model	chinese-electra-180g-base-discriminator	相同
> loss_type	ce	ce
> err_weight_mult	2.0	2.0
> final_class_weights[ERR]	11.529	11.573
> val_ratio	0.05	0.05
> train_n	1,908,531	1,908,150
> val_n	100,450	100,429
> metric_for_best_model	eval_recall	eval_recall
> epochs	3	3
> batch_size	8	8
> lr	2e-5	2e-5
> seed	42	42
> precision	fp32	fp32（推测）
> ### 评测结果对比
> 评测集：`val_task_stratified_char.jsonl`（7500 条，spelling / grammar / mixed 各 2500）
> 
> #### 字级整体
> 指标	本次训练	原始模型	差值
> Accuracy	0.9461	0.9081	+0.038
> Precision (ERR)	0.3637	0.2562	+0.107
> **Recall (ERR)**	**0.6745**	**0.8222**	**-0.148 ↓**
> F1 (ERR)	0.4726	0.3906	+0.082
> #### 字级按任务拆分
> Task	指标	本次训练	原始模型	差值
> spelling	Recall	0.979	0.986	-0.007
> grammar	Recall	0.683	0.823	-0.140 ↓
> mixed	Recall	0.501	0.732	-0.231 ↓
> #### 句级（最优阈值）
> 指标	本次训练（thr=0.65）	原始模型（thr=0.70）
> Precision	81.8%	78.7%
> Recall	91.0%	94.6%
> F1	**86.1%**	**85.9%**
> 句级 F1 基本持平（差 0.2%），但字级 ERR Recall 差距集中在 grammar 和 mixed 两类。
> 
> ### 问题
> 我目前使用的是 `train_stage1.json` + `train_stage2.json` 直接合并得到的 `train_stage_all.jsonl`（共 2,008,981 条，训练集 1,908,531 条）。
> 
> 请问原始训练使用的 `cec_train.jsonl` 是如何从 `train_stage1/train_stage2` 生成的？
> 
> 具体想确认：
> 
> 1. 训练集是否经过了额外的清洗、去重或过滤步骤？
> 2. 生成脚本是否可以公开？
> 3. 请问是否我还有其他遗漏的超参？
> 
> 非常感谢！！！

原始模型是指我训好发布的那个？

**Comment by crisyy** [comment]
是的，就是尝试能还原您发布的ChineseErrorDetectorElectra模型的效果呢，我在您ChineseErrorDetectorElectra 模型基础上也尝试做了几次全参的训练，但是Recall的结果一直达不到您的效果。

**Comment by Shixurong** [comment]
好的，稍等我整理一下发你

Sent&nbsp;from&nbsp;my&nbsp;iPhone


------------------ 原始邮件 ------------------
发件人: crisyy ***@***.***&gt;
发送时间: 2026年6月24日 15:32
收件人: TW-NLP/ChineseErrorCorrector ***@***.***&gt;
抄送: littlebee ***@***.***&gt;, Comment ***@***.***&gt;
主题: Re: [TW-NLP/ChineseErrorCorrector] 【请教/咨询】关于字级检错门控模型 ChineseErrorDetectorElectra 训练源码与细节的咨询 (Issue #49)



crisyy left a comment (TW-NLP/ChineseErrorCorrector#49)
 
是的，就是尝试能还原您发布的ChineseErrorDetectorElectra模型的效果呢，我在您ChineseErrorDetectorElectra 模型基础上也尝试做了几次全参的训练，但是Recall的结果一直达不到您的效果。
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by Shixurong** [comment]
> 是的，就是尝试能还原您发布的ChineseErrorDetectorElectra模型的效果呢，我在您ChineseErrorDetectorElectra 模型基础上也尝试做了几次全参的训练，但是Recall的结果一直达不到您的效果。

不好意思最近太忙了
[chinese_sentence_error_detector_repro_scripts.tar.gz](https://github.com/user-attachments/files/31056307/chinese_sentence_error_detector_repro_scripts.tar.gz)


感谢反馈。关于 `cec_train.jsonl` 的生成与复现，说明如下。

### 1. `cec_train.jsonl` 如何从 train_stage1 / train_stage2 得到？

不是直接把两个 JSON 文件简单 cat 合并。正确路径是读取 Hugging Face 数据集 [`twnlp/ChinseseErrorCorrectData`](https://huggingface.co/datasets/twnlp/ChinseseErrorCorrectData) 的 **`train` split**（该 split 已包含 stage1 + stage2），再用脚本转成每行 `{"source","target"}`：

```bash
python prepare_chinsese_errorcorrectdata_jsonl.py \
  --from_hub \
  --split train \
  --out_jsonl data/cec_train.jsonl
```

脚本逻辑：从 `conversations` 中取 human/gpt，从 human 侧截取「输入句子为：」后的原文作为 `source`，gpt 侧作为 `target`。

若本地已有 `save_to_disk` 目录，可用 `--from_disk ... --split train`。

上面提到的「合并 stage1+stage2 约 200 万条」在**条数量级**上是对的；但请确认已按上述规则转成 `source`/`target`，而不是保留原始对话格式。

### 2. 训练集是否经过额外清洗、去重或过滤？

**没有。** 转换脚本仅跳过解析失败的样本（缺 conversations / 抽不出 source 或 target）。  
训练时 `train_char.py` 只会再过滤「归一化后为空」的文本。  
**没有**额外语义清洗、去重、长度过滤等步骤。

### 3. 生成脚本是否可以公开？

可以。相关脚本已整理进附件压缩包（或仓库内同名文件），主要包括：

- `prepare_chinsese_errorcorrectdata_jsonl.py`（CEC → JSONL）
- `char_align.py` / `normalize.py` / `prepare_char_jsonl.py`（字级标签）
- `train_char.py` / `evaluate_char.py`
- `TRAIN_CHAR_ELECTRA.txt`、`training_char_meta.json`（本次 ELECTRA 训练实际超参记录）

### 4. 是否还有遗漏的超参？

请优先对齐我们实际 checkpoint 的 meta（`training_char_meta.json`）：

| 项 | 值 |
|----|-----|
| 基座 | `hfl/chinese-electra-180g-base-discriminator` |
| 数据 | `--jsonl data/cec_train.jsonl` |
| `val_jsonl` | **未使用**（`null`） |
| `val_ratio` | `0.05`（从 train 内随机切 5% 做 early stopping） |
| `seed` | `42` |
| `max_length` | `256` |
| `epochs` | `3` |
| `batch_size` | `8` |
| `lr` | `2e-5` |




| `warmup_ratio` | `0.06` |
| `weight_decay` | `0.01` |
| `metric_for_best` | **`recall`** |
| `loss_type` | `ce` |
| `class_weight_cap` | `10.0`（自动统计类权重后再 cap） |
| `err_weight_mult` | **`2.0`**（容易漏） |
| `neg_subsample_keep` | `1.0` |
| 混合精度 | `--fp16` |

您看到的「全量约 200 万 → 训练集约 190 万」对应的就是默认 `val_ratio=0.05`，不是额外过滤掉了约 10 万条。

另外请注意评测集对齐：

- 训练时内部验证：从 `cec_train` 随机 5%
- 报表常用：`val_task_stratified.jsonl`（需单独用 `prepare_task_validation_jsonl.py` 生成）
- 官方 hold-out：`cec_validation.jsonl`（Hub `validation` split）

不同验证集上的指标不可直接对比；句级门控还需对齐 `evaluate_char.py --sweep_sentence` 得到的 `sentence_threshold`。

### 5. 建议复现命令（ELECTRA）

```bash
python prepare_chinsese_errorcorrectdata_jsonl.py --from_hub --split train --out_jsonl data/cec_train.jsonl

python train_char.py \
  --model_name_or_path hfl/chinese-electra-180g-base-discriminator \
  --jsonl data/cec_train.jsonl \
  --output_dir outputs/char_detector_electra \
  --fp16 --batch_size 8 --epochs 3 \
  --metric_for_best recall \
  --err_weight_mult 2.0
```

如仍有约 15% 差距，请补充：评测用的 JSONL 路径、字级/句级指标、以及 `sentence_threshold`，我们再一起对一下。
